test/e2e/daemon.go

```diff
@@ -0,0 +1,235 @@
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Daemon", func() {
+	var c *client.Client
+
+	BeforeEach(func() {
+		var err error
+		c, err = loadClient()
+		Expect(err).NotTo(HaveOccurred())
+		err = clearNodeLabels(c)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		err := clearNodeLabels(c)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should launch a daemon pod on every node of the cluster", func() {
+		testDaemons(c)
+	})
+})
+
+func clearNodeLabels(c *client.Client) error {
+	nodeClient := c.Nodes()
+	nodeList, err := nodeClient.List(labels.Everything(), fields.Everything())
+	if err != nil {
+		return err
+	}
+	for _, node := range nodeList.Items {
+		if len(node.Labels) != 0 {
+			node.Labels = map[string]string{}
+			newNode, err := nodeClient.Update(&node)
+			if err != nil {
+				return err
+			} else if len(newNode.Labels) != 0 {
+				return fmt.Errorf("Could not make node labels nil.")
+			}
+		}
+	}
+	return nil
+}
+
+func checkDaemonPodOnNodes(c *client.Client, selector map[string]string, nodeNames []string) func() (bool, error) {
+	// Don't return an error, because returning an error will abort wait.Poll, but
+	// if there's an error, we want to try getting the daemon again.
+	return func() (bool, error) {
+		// Get list of pods satisfying selector.
+		podList, err := c.Pods(api.NamespaceDefault).List(labels.Set(selector).AsSelector(), fields.Everything())
+		if err != nil {
+			return false, nil
+		}
+		pods := podList.Items
+
+		// Get a map of node names to number of daemon pods running on the node.
+		nodesToPodCount := make(map[string]int)
+		for _, pod := range pods {
+			nodesToPodCount[pod.Spec.NodeName] += 1
+		}
+
+		// Ensure that exactly 1 pod is running on all nodes in nodeNames.
+		for _, nodeName := range nodeNames {
+			if nodesToPodCount[nodeName] != 1 {
+				return false, nil
+			}
+		}
+
+		// Ensure that sizes of the lists are the same. We've verified that every element of nodeNames is in
+		// nodesToPodCount, so verifying the lengths are equal ensures that there aren't pods running on any
+		// other nodes.
+		return len(nodesToPodCount) == len(nodeNames), nil
+	}
+}
+
+func checkRunningOnAllNodes(c *client.Client, selector map[string]string) func() (bool, error) {
+	// Don't return an error, because returning an error will abort wait.Poll, but
+	// if there's an error, we want to try getting the daemon again.
+	return func() (bool, error) {
+		nodeList, err := c.Nodes().List(labels.Everything(), fields.Everything())
+		if err != nil {
+			return false, nil
+		}
+		nodeNames := make([]string, 0)
+		for _, node := range nodeList.Items {
+			nodeNames = append(nodeNames, node.Name)
+		}
+		return checkDaemonPodOnNodes(c, selector, nodeNames)()
+	}
+}
+
+func checkRunningOnNoNodes(c *client.Client, selector map[string]string) func() (bool, error) {
+	return checkDaemonPodOnNodes(c, selector, make([]string, 0))
+}
+
+func testDaemons(c *client.Client) {
```

Comment:
I agree that the function can be broken up :)

However, I think we should not break it up into 2 disjoint tests. We should test that 2 daemon controllers can co-exist together (as long as their selectors are different).